flusher.go

```diff
@@ -82,85 +143,192 @@ func (s *Server) Flush(interval time.Duration, metricLimit int) {
+	jmLength := 0
+	for _, wm := range wms {
+		jmLength += len(wm.histograms)
+		jmLength += len(wm.sets)
+		jmLength += len(wm.timers)
+	}
+
+	jsonMetrics := make([]JSONMetric, 0, jmLength)
+	for _, wm := range wms {
+		for _, histo := range wm.histograms {
+			jm, err := histo.Export()
+			if err != nil {
+				s.logger.WithFields(logrus.Fields{
+					logrus.ErrorKey: err,
+					"type":          "histogram",
+					"name":          histo.name,
+				}).Error("Could not export metric")
+				continue
+			}
+			jsonMetrics = append(jsonMetrics, jm)
+		}
+		for _, set := range wm.sets {
+			jm, err := set.Export()
+			if err != nil {
+				s.logger.WithFields(logrus.Fields{
+					logrus.ErrorKey: err,
+					"type":          "set",
+					"name":          set.name,
+				}).Error("Could not export metric")
+				continue
+			}
+			jsonMetrics = append(jsonMetrics, jm)
+		}
+		for _, timer := range wm.timers {
+			jm, err := timer.Export()
+			if err != nil {
+				s.logger.WithFields(logrus.Fields{
+					logrus.ErrorKey: err,
+					"type":          "timer",
+					"name":          timer.name,
+				}).Error("Could not export metric")
+				continue
+			}
+			jsonMetrics = append(jsonMetrics, jm)
+		}
+	}
+
+	s.statsd.Gauge("forward.post_metrics_total", float64(len(jsonMetrics)), nil, 1.0)
+	if len(jsonMetrics) == 0 {
+		s.logger.Info("Nothing to forward, skipping.")
 		return
 	}
-	// make sure to flush remaining compressed bytes to the buffer
-	compressor.Close()
-	s.statsd.TimeInMilliseconds(
-		"flush.part_duration_ns",
-		float64(time.Now().Sub(cstart).Nanoseconds()),
-		[]string{"part:marshal"},
-		1.0,
-	)
-	// Len reports the unread length, so we have to record this before it's POSTed
-	bodyLength := reqBody.Len()
-	s.statsd.Histogram("flush.content_length_bytes", float64(bodyLength), nil, 1.0)
 
-	req, err := http.NewRequest(http.MethodPost, fmt.Sprintf("%s/api/v1/series?api_key=%s", s.DDHostname, s.DDAPIKey), &reqBody)
+	endpoint, err := resolveEndpoint(fmt.Sprintf("%s/import", s.ForwardAddr))
 	if err != nil {
-		s.statsd.Count("flush.error_total", 1, []string{"cause:construct"}, 1.0)
-		s.logger.WithError(err).Error("Error constructing POST request")
-		return
+		// not a fatal error, we'll just try to use the host as it was given to us
+		s.statsd.Count("forward.error_total", 1, []string{"cause:dns"}, 1.0)
+		s.logger.WithError(err).Warn("Could not re-resolve host for forward")
+	}
+
+	// the error has already been logged (if there was one), so we only care
+	// about the success case
+	if s.postHelper(endpoint, jsonMetrics, "forward") == nil {
+		s.logger.WithField("metrics", len(jsonMetrics)).Info("Completed forward to upstream Veneur")
+	}
+}
+
+// given a url, attempts to resolve the url's host, and returns a new url whose
+// host has been replaced by the first resolved address
+// on failure, it returns the argument, and the resulting error
+func resolveEndpoint(endpoint string) (string, error) {
```

Comment:
correct, if we don't perform the resolution ourselves then we keep getting the cached host for at least several minutes after failing over the DNS record.